legion/engine/llri-vk/detail/command_group.cpp

```diff
@@ -0,0 +1,130 @@
+/**
+ * @file command_group.cpp
+ * @copyright 2021-2021 Leon Brands. All rights served.
+ * @license: https://github.com/Legion-Engine/Legion-LLRI/blob/main/LICENSE
+ */
+
+#include <llri/llri.hpp>
+#include <llri-vk/utils.hpp>
+
+namespace LLRI_NAMESPACE
+{
+    result CommandGroup::impl_reset()
+    {
+        const auto r = static_cast<VolkDeviceTable*>(m_deviceFunctionTable)->
+            vkResetCommandPool(static_cast<VkDevice>(m_device->m_ptr), static_cast<VkCommandPool>(m_ptr), VK_COMMAND_POOL_RESET_RELEASE_RESOURCES_BIT);
+
+        if (r != VK_SUCCESS)
+            return internal::mapVkResult(r);
+
+        for (auto* cmdList : m_cmdLists)
+            cmdList->m_state = command_list_state::Empty;
+
+        return result::Success;
+    }
+
+    result CommandGroup::impl_allocate(const command_list_alloc_desc& desc, CommandList** cmdList)
+    {
+        VkCommandBufferAllocateInfo allocInfo { VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO, nullptr };
+        allocInfo.commandPool = static_cast<VkCommandPool>(m_ptr);
+        allocInfo.commandBufferCount = 1;
+        allocInfo.level = internal::mapCommandListUsage(desc.usage);
+
+        VkCommandBuffer cmd;
+        auto r = static_cast<VolkDeviceTable*>(m_deviceFunctionTable)->vkAllocateCommandBuffers(static_cast<VkDevice>(m_device->m_ptr), &allocInfo, &cmd);
+        if (r != VK_SUCCESS)
+            return internal::mapVkResult(r);
+
+        auto* output = new CommandList();
+        output->m_ptr = cmd;
+        output->m_group = this;
+
+        output->m_deviceHandle = m_device->m_ptr;
+        output->m_deviceFunctionTable = m_deviceFunctionTable;
+
+        output->m_nodeMask = desc.nodeMask;
+        output->m_usage = desc.usage;
+        output->m_state = command_list_state::Empty;
+
+        output->m_validationCallback = m_validationCallback;
+        output->m_validationCallbackMessenger = m_validationCallbackMessenger;
+
+        m_cmdLists.emplace(output);
+
+        *cmdList = output;
+        return result::Success;
+    }
+
+    result CommandGroup::impl_allocate(const command_list_alloc_desc& desc, uint8_t count, std::vector<CommandList*>* cmdLists)
+    {
+        VkCommandBufferAllocateInfo allocInfo { VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO, nullptr };
+        allocInfo.commandPool = static_cast<VkCommandPool>(m_ptr);
+        allocInfo.commandBufferCount = count;
+        allocInfo.level = internal::mapCommandListUsage(desc.usage);
+
+        std::vector<VkCommandBuffer> cmdBuffers(count);
+        auto r = static_cast<VolkDeviceTable*>(m_deviceFunctionTable)->vkAllocateCommandBuffers(static_cast<VkDevice>(m_device->m_ptr), &allocInfo, cmdBuffers.data());
+        if (r != VK_SUCCESS)
+            return internal::mapVkResult(r);
+
+        for (uint8_t i = 0; i < count; i++)
+        {
+            auto* cmdList = new CommandList();
+            cmdList->m_ptr = cmdBuffers[i];
```

Comment:
I get trying to use a uint8_t to save on a bit of memory, but you are slowing down this loop with it. vectors and arrays all use size_t and pointers use ptrdiff_t. rn it needs to implicitly upcast the uint8_t to a size_t with every iteration.